index.js

```diff
@@ -0,0 +1,153 @@
+  let { username, password } = req.body;
+  db.findBy({ username })
+    .first()
+    .then(user => {
+      // if user exists and password is the password put in for login and user.password is the hashed
+      //  password which was converted into hashed at the time of register.bcrypt is written in
+      //  such a way that it can compare thre original woth the hashed one.
+      if (user && bcrypt.compareSync(password, user.password)) {
+        req.session.user = user;
+        console.log("Request Session", req.session);
+        res.status(200).json({
+          message: `Welcome ${
+            user.username
+          }!,you are logged in successfully,have a cookie`
+        });
+      } else {
+        res.status(401).json({ message: "You can't pass" });
+      }
+    })
+    .catch(error => {
+      res.status(500).json(error);
+    });
+});
+//now we want to go in a restrictred route  which can only be acccessed if we have a session and a
+//session user,so commenting out everything below because we have already verified in login that we
+// have a valid user having username and password.in login, the only way we can have req.session.user
+// is if that user has been validated.we don't have to check every single time.so, now we have to check whether
+//a session exist or not, we can find on request.what would be true ..if the user is logged inwe have asession and username on the session
+//if we are in the same domain localhost 5000, we have access to the cookie
+// function restricted (req,res,next){
+//   console.log("Req : ", req)
+//    const {username,password}=req.headers;
+//    if(username && password){
+//       db.findBy({username})
+//         .first()
+//         .then(user =>{
+//             if(user && bcrypt.compareSync(password,user.password)){
+//              next()
+//             }else{
+//               res.status(401).json({message:"you can't pass"})
+
+//             }
+//         })
+//         .catch(error =>{
+//           res.status(500).json(error);
+//         })
+
+//    }else{
+//     res.status(400).json({message:"No credentials provided"})
+
+//    }
+// }
+
+function restricted(req, res, next) {
+  console.log(req.session);
+  if (req.session && req.session.user) {
+    next();
+  } else {
+    res
+      .status(401)
+      .json({ message: "you shall not pass or show him the login page" });
+  }
+}
+
+//protect this route only authenticated  users can see it.
+server.get("/api/users", restricted, (req, res) => {
+  console.log("Get users");
+  db.find()
+    .then(users => {
+      res.json(users);
+    })
+    .catch(err => res.send(err));
+});
+
+server.get("/api/logout", (req, res) => {
+  if (req.session) {
+    req.session.destroy(err => {
+      //req.session.destroy takes a callback
+      if (err) {
+        res.status(400).json({ message: "you can't leave the session" });
+      } else {
+        res.send("bye,bye");
+      }
+    });
+  } else {
+    res.send();
```

Comment:
Some sort of message here might be preferred (i.e., "logout attempted, but no session found"); this might be grounds for sending an error back since the route wasn't used appropriately in this case.